src/python/WMArchive/PySpark/RecordAggregator.py

```diff
@@ -14,71 +27,125 @@ def mapper(self, records):
+
+                # Performance
+                if not stats.get('performance'):
+                    stats['performance'] = {
+                        "cpu": {
+                            "TotalJobCPU": 0,
+                            "TotalJobTime": 0,
+                        },
+                        "storage": {
+                            "readTotalMB": 0,
+                            "writeTotalMB": 0,
+                        }
+                    }
+                performance_cpu = step['performance']['cpu']
+                if performance_cpu.get('TotalJobCPU', 0):
+                    stats['performance']['cpu']['TotalJobCPU'] += performance_cpu.get('TotalJobCPU', 0)
+                if performance_cpu.get('TotalJobTime', 0):
+                    stats['performance']['cpu']['TotalJobTime'] += performance_cpu.get('TotalJobTime', 0)
+                performance_storage = step['performance']['storage']
+                if performance_storage.get('readTotalMB', 0):
+                    stats['performance']['storage']['readTotalMB'] += performance_storage.get('readTotalMB', 0)
+                if performance_storage.get('writeTotalMB', 0):
+                    stats['performance']['storage']['writeTotalMB'] += performance_storage.get('writeTotalMB', 0)
+
+                # Store stats in document
+                document['stats'][scope_hash] = stats
+
+        return document
 
     def reducer(self, records, init=0):
         "Simpler reducer which collects all results from RDD.collect() records"
-        out = {}
-        count = 0
-        mdict = {'cpu':0, 'time':0, 'rsize':0, 'wsize':0, 'doc':0}
-        for rec in records:
-            for host, hdict in rec.items():
-                if  host not in out.keys():
-                    out[host] = {}
-                _hdict = out[host]
-                for site, sdict in hdict.items():
-                    if  site not in _hdict.keys():
-                        _hdict = {site:mdict}
-                        out[host][site] = mdict
-                    _sdict = _hdict[site]
-                    _cpu = _sdict['cpu'] + sdict['cpu']
-                    _time = _sdict['time'] + sdict['time']
-                    _rsize = _sdict['rsize'] + sdict['rsize']
-                    _wsize = _sdict['wsize'] + sdict['wsize']
-                    _doc = _sdict['doc'] + sdict['doc']
-                    ndict = {'cpu':_cpu, 'time':_time, 'rsize':_rsize, 'wsize':_wsize, 'doc':_doc}
-                    out[host][site] = ndict
-            count += 1
-	# simple printout of summary info
-        for host, hdict in out.items():
-            print('\n### %s' % host)
-            for site, sdict in hdict.items():
-                print('')
-                print(site)
-                for key, val in sdict.items():
-                    print('%s: %s' % (key, val))
-        return out
+        document = {
+            'stats': {},
+        }
+        for existing_document in records:
```

Comment:
This is code will perform poorly CPU/RAM wise since you're creating a data structure document (which is a dict), then you write it to file and then insert a giganic dict into MongoDB. Instead, I'll suggest to refactor this code and create a function which will yield documents. This generator can be passed to MongoDB client which will insert docs into MongoDB (but you'll need to use proper chunk_size). Let me know if you understand how to refactor this piece.